sdk/go/common/util/cmdutil/exit.go

```diff
@@ -125,6 +126,32 @@ func RunFunc(run func(cmd *cobra.Command, args []string) error) func(*cobra.Comm
 	}
 }
 
+// Execute runs the given command, handling any errors that occur.  This should be used at the top level
+// of a Pulumi command to ensure that any errors are handled correctly.
+func Execute(ctx context.Context, cmd *cobra.Command) {
+	err := cmd.ExecuteContext(ctx)
+	if err != nil {
+		// If we were asked to bail, that means we already printed out a message.  We just need
+		// to quit at this point (with an error code so no one thinks we succeeded).  Bailing
+		// always indicates a failure, just one we don't need to print a message for.
+		if result.IsBail(err) {
+			os.Exit(-1)
+			return
+		}
+
+		// If there is a stack trace, and logging is enabled, append it.  Otherwise, debug logging it.
+		var msg string
+		if logging.LogToStderr {
+			msg = DetailedError(err)
+		} else {
+			msg = errorMessage(err)
+			logging.V(3).Info(DetailedError(err))
+		}
+
+		ExitError(msg)
```

Comment:
I think we need to pull `processCmdErrors` (originally called by `runCmdFunc`) up to this function, so that we e.g. spit out panic messages for snapshot errors and help texts for decryption issues. But I agree this is a nicer approach, and equally we can probably work to remove all this stuff in `cmdutil` (and when we kill the other stuff in `cmdutil`, the package in its entirety)

Reply:
Sigh looking into this cobra is just a hot mess for error handling, I'm going to revert these cmd changes for now and just stick to the separate struct method we've used so far.

Reply:
Reworked to just use the struct style for now. Cobra will need more looking into.